client/consensus-spin/src/standalone.rs

```diff
@@ -0,0 +1,440 @@
+// Copyright (C) Quantum Fusion Network, 2025.
+// Copyright (C) Parity Technologies (UK) Ltd., until 2025.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Standalone functions used within the implementation of SPIN.
+
+use std::fmt::Debug;
+
+use log::trace;
+
+use codec::Codec;
+
+use sc_client_api::UsageProvider;
+use sp_api::{Core, ProvideRuntimeApi};
+use sp_application_crypto::{AppCrypto, AppPublic};
+use sp_blockchain::Result as CResult;
+use sp_consensus::Error as ConsensusError;
+use sp_consensus_slots::Slot;
+use sp_core::crypto::{ByteArray, Pair};
+use sp_keystore::KeystorePtr;
+use sp_runtime::{
+    DigestItem,
+    traits::{Block as BlockT, Header, NumberFor, Zero},
+};
+
+pub use sc_consensus_slots::check_equivocation;
+
+use super::{
+    AuthorityId, CompatibilityMode, CompatibleDigestItem, LOG_TARGET, SessionLength, SlotDuration,
+    SpinApi, SpinAuxData,
+};
+
+/// Get the slot duration for SPIN by reading from a runtime API at the best block's state.
+pub fn slot_duration<A, B, C>(client: &C) -> CResult<SlotDuration>
+where
+    A: Codec,
+    B: BlockT,
+    C: ProvideRuntimeApi<B> + UsageProvider<B>,
+    C::Api: SpinApi<B, A>,
+{
+    slot_duration_at(client, client.usage_info().chain.best_hash)
+}
+
+/// Get the slot duration for SPIN by reading from a runtime API at a given block's state.
+pub fn slot_duration_at<A, B, C>(client: &C, block_hash: B::Hash) -> CResult<SlotDuration>
+where
+    A: Codec,
+    B: BlockT,
+    C: ProvideRuntimeApi<B>,
+    C::Api: SpinApi<B, A>,
+{
+    client
+        .runtime_api()
+        .slot_duration(block_hash)
+        .map_err(|err| err.into())
+}
+
+/// Get the slot author for given block along with authorities.
+///
+/// Session increment is assumed to be done by the runtime
+pub fn slot_author<P: Pair>(
+    slot: Slot,
+    session_length: SessionLength,
+    authorities: &[AuthorityId<P>],
+) -> Option<&AuthorityId<P>> {
+    if authorities.is_empty() {
+        return None;
+    }
+    assert!(session_length != 0, "session_length can not be zero; qed");
```

Comment:
Let's use [`assert_ne!()`](https://doc.rust-lang.org/std/macro.assert_ne.html)
```suggestion
    assert_ne!(session_length, 0, "session_length can not be zero; qed");
```

Reply:
why? imo, it looks less readable, don't see any added benefit

Reply:
I would prefer:
```rust
assert_eq!(a, b);
<over>
assert!(a == b);
```
because `assert_eq!()` prints values if they are not equal:
```
thread 'main' panicked at src/main.rs:2:5:
assertion `left == right` failed
  left: 1
 right: 2
``` 

And for `assert_ne!()` it's the same if we compare two variables
```rust
assert_ne!(a, b);
```
Let's use `assert_eq` and `assert_ne` for consistency.

Reply:
Ok, looks like it is a bit cleaner